### PR TITLE
OCPNODE-367: Auto node sizing for Openshift

### PR DESCRIPTION
**Pull request by harche**
Signed-off-by: Harshal Patil <harpatil@redhat.com>

**- What I did**
Introduce auto node sizing for Openshift - https://issues.redhat.com/browse/OCPNODE-367

**- How to verify it**
To enable auto node sizing on worker node use following yaml

```yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: KubeletConfig
metadata:
  name: dynamic-node 
spec:
  autoSizingReserved: true 
  machineConfigPoolSelector:
    matchLabels:
      pools.operator.machineconfiguration.openshift.io/worker: ""
```
Once it's rolled out, log into the worker node,

```bash
# ps -ef | grep kubelet
root        1613       1 11 06:49 ?        00:00:05 kubelet --config=/etc/kubernetes/kubelet.conf --bootstrap-kubeconfig=/etc/kubernetes/kubeconfig --kubeconfig=/var/lib/kubelet/kubeconfig --container-runtime=remote --container-runtime-endpoint=/var/run/crio/crio.sock --runtime-cgroups=/system.slice/crio.service --node-labels=node-role.kubernetes.io/worker,node.openshift.io/os_id=rhcos --node-ip= --minimum-container-ttl-duration=6m0s --volume-plugin-dir=/etc/kubernetes/kubelet-plugins/volume/exec --cloud-provider=azure --cloud-config=/etc/kubernetes/cloud.conf --pod-infra-container-image=quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:7b8e2e2857d8ac3499c9eb4e449cc3296409f1da21aa21d0140134d611e65b84 --system-reserved=cpu=0.07,memory=2.5Gi --v=2
```
You should be able to see kubelet's `--system-reserved` flag is set for optimal values of `cpu` and `memory`

**- Description for the changelog**
Enable auto node sizing post installation

### CODE REVIEW

File: pkg/controller/kubelet-config/helpers.go [hunk 1]
```diff
@@ -23,6 +23,29 @@ import (
...
+
+	mode := 0644
+	overwrite := true
+	du := dataurl.New([]byte(config), "text/plain")
+	du.Encoding = dataurl.EncodingASCII
+	duStr := du.String()
+
+	return &ign3types.File{
+		Node: ign3types.Node{
+			Path:      "/etc/node-sizing-enabled.env",
```
Comment (rphillips) [inline pkg/controller/kubelet-config/helpers.go:37]: this should be in /etc/kubernetes

**Comment by harche** [comment]
Tested these changes on `Azure` using following sample pod,

```yaml
apiVersion: v1
kind: ReplicationController
metadata:
  name: badmem
  namespace: default
spec:
  replicas: 1
  selector:
    app: badmem
  template:
    metadata:
      labels:
        app: badmem
    spec:
      containers:
      - args:
        - python
        - -c
        - |
          x = []
          while True:
            x.append("x" * 1048576)
        image: registry.redhat.io/rhel7:latest
        name: badmem
```
The node never went in `NotReady` state. 
```bash
$ oc get nodes 
NAME                                                STATUS                     ROLES    AGE    VERSION
ci-ln-3z6xtwk-002ac-c4l9j-master-0                  Ready                      master   139m   v1.20.0+5fbfd19
ci-ln-3z6xtwk-002ac-c4l9j-master-1                  Ready                      master   139m   v1.20.0+5fbfd19
ci-ln-3z6xtwk-002ac-c4l9j-master-2                  Ready                      master   139m   v1.20.0+5fbfd19
ci-ln-3z6xtwk-002ac-c4l9j-worker-centralus1-pbxb6   Ready,SchedulingDisabled   worker   130m   v1.20.0+5fbfd19
ci-ln-3z6xtwk-002ac-c4l9j-worker-centralus2-k67s7   Ready                      worker   130m   v1.20.0+5fbfd19
ci-ln-3z6xtwk-002ac-c4l9j-worker-centralus3-fqghh   Ready,SchedulingDisabled   worker   130m   v1.20.0+5fbfd19
```
and the pods were getting evicted pretty smoothly,
```bash
$ oc get pods 
NAME           READY   STATUS    RESTARTS   AGE
badmem-2cxws   0/1     Evicted   0          13s
badmem-2grjc   0/1     Evicted   0          15s
badmem-2j6vq   0/1     Evicted   0          13s
badmem-2knnz   0/1     Evicted   0          12s
badmem-2rplp   0/1     Evicted   0          14s
badmem-45njp   0/1     Evicted   0          13s
badmem-4hnzt   0/1     Evicted   0          14s
badmem-55p9v   0/1     Evicted   0          14s
badmem-56hcd   0/1     Evicted   0          13s
badmem-58zpx   0/1     Evicted   0          12s
badmem-68tls   0/1     Evicted   0          12s
badmem-6b54b   0/1     Evicted   0          13s
badmem-74phw   0/1     Evicted   0          14s
badmem-76hzv   0/1     Evicted   0          12s
badmem-7dsk9   0/1     Evicted   0          13s
badmem-7ncj6   0/1     Evicted   0          14s
badmem-8kv79   0/1     Evicted   0          13s
badmem-8mclq   0/1     Evicted   0          13s
badmem-8rv8m   0/1     Evicted   0          14s
badmem-8z5gk   0/1     Evicted   0          12s
```

/cc @rphillips

**Comment by harche** [comment]
I forgot to address https://github.com/openshift/machine-config-operator/pull/2466#discussion_r597714810, will remove the hold after making the suggested changes.

File: pkg/controller/kubelet-config/helpers.go (hunk 1 above)
Comment (harche) [reply pkg/controller/kubelet-config/helpers.go:37]: When the file `node-sizing-enabled.env` is moved in /etc/kubernetes, it gets `system_u:object_r:kubernetes_file_t:s0` SELinux labels. For some reason, the system unit file isn't able to import this file,

```
# systemctl status kubelet-auto-node-size.service
● kubelet-auto-node-size.service - Dynamically sets the system reserved for the kubelet
   Loaded: loaded (/etc/systemd/system/kubelet-auto-node-size.service; enabled; vendor preset: disabled)
   Active: failed (Result: resources)
      CPU: 0

Mar 22 11:35:11 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: kubelet-auto-node-size.service: Failed to load environment files: Permission denied
Mar 22 11:35:11 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: kubelet-auto-node-size.service: Failed to run 'start' task: Permission denied
Mar 22 11:35:11 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: kubelet-auto-node-size.service: Failed with result 'resources'.
Mar 22 11:35:11 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: Failed to start Dynamically sets the system reserved for the kubelet.
Mar 22 11:41:38 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: kubelet-auto-node-size.service: Failed to load environment files: Permission denied
Mar 22 11:41:38 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: kubelet-auto-node-size.service: Failed to run 'start' task: Permission denied
Mar 22 11:41:38 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: kubelet-auto-node-size.service: Failed with result 'resources'.
Mar 22 11:41:38 testclustllb1-hltqw-worker-a-klvm6 systemd[1]: Failed to start Dynamically sets the system reserved for the kubelet.
```

Looking into it.

Comment (harche) [reply pkg/controller/kubelet-config/helpers.go:37]: `aureport` confirms it,

```
12. 03/22/2021 11:35:11 systemd system_u:system_r:init_t:s0 0 file read system_u:object_r:kubernetes_file_t:s0 denied 45
13. 03/22/2021 11:41:38 systemd system_u:system_r:init_t:s0 0 file read system_u:object_r:kubernetes_file_t:s0 denied 86
```

Comment (harche) [reply pkg/controller/kubelet-config/helpers.go:37]: That env file is created on first boot using, https://github.com/openshift/machine-config-operator/pull/2466/files#diff-c56bef4bc0ff73a538d970d1b4d9684b3272703584bbfe8002dc9f5c81368adfR1

Comment (harche) [reply pkg/controller/kubelet-config/helpers.go:37]: I haven't updated the changes that move that env file from `/etc/` to `/etc/kubernetes` in github. The failures that I have pasted above are from my local setup after making the changes. So in this PR you are still seeing `/etc/` paths and not `/etc/kubernetes`

**Comment by rphillips** [comment]
lgtm

File: templates/common/_base/files/kubelet-auto-sizing.yaml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+contents:
+  inline: |
+    #!/bin/bash
+    set -e
+    NODE_SIZES_ENV=${NODE_SIZES_ENV:-/etc/node-sizing.env}
+    function dynamic_memory_sizing {
+        total_memory=$(free -g|awk '/^Mem:/{print $2}')
+        # total_memory=8 test the recommended values by modifying this value
+        recommended_systemreserved_memory=0
+        if (($total_memory <= 4)); then # 25% of the first 4GB of memory
```
Comment (sinnykumari) [inline templates/common/_base/files/kubelet-auto-sizing.yaml:12]: Logic for `dynamic_memory_sizing` func seems a bit unreadable. Like condition `if (($total_memory <= 4))` has been used twice. Also, multiple if condition will be called based on what memory size for e.g. 2 G memory. Is this logic coming already testing in upstream?

File: templates/common/_base/files/kubelet-auto-sizing.yaml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        else
+            recommended_systemreserved_memory=$(echo $recommended_systemreserved_memory 6.72 | awk '{print $1 + $2}')
+            total_memory=$((total_memory-112))
+        fi
+        if (($total_memory >= 128)); then # 2% of any memory above 128GB
+            recommended_systemreserved_memory=$(echo $recommended_systemreserved_memory $(echo $total_memory 0.02 | awk '{print $1 * $2}') | awk '{print $1 + $2}')
+        fi
+        echo "SYSTEM_RESERVED_MEMORY=${recommended_systemreserved_memory}Gi">> ${NODE_SIZES_ENV}
+    }
+    function dynamic_cpu_sizing {
```
Comment (sinnykumari) [inline templates/common/_base/files/kubelet-auto-sizing.yaml:45]: Same concern as above.

**Review by sinnykumari — COMMENTED** [review]
I think we will also need to update kubelet.service for on-perm https://github.com/openshift/machine-config-operator/blob/master/templates/master/01-master-kubelet/on-prem/units/kubelet.service.yaml and https://github.com/openshift/machine-config-operator/blob/master/templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml

File: templates/common/_base/files/kubelet-auto-sizing.yaml (hunk 2 above)
Comment (sinnykumari) [reply templates/common/_base/files/kubelet-auto-sizing.yaml:12]: Talked offline to Harshal and it seems this is the correct approach.

**Comment by harche** [comment]
> I think we will also need to update kubelet.service for on-perm https://github.com/openshift/machine-config-operator/blob/master/templates/master/01-master-kubelet/on-prem/units/kubelet.service.yaml and https://github.com/openshift/machine-config-operator/blob/master/templates/worker/01-worker-kubelet/on-prem/units/kubelet.service.yaml

Good catch @sinnykumari, thanks for pointing that out. I have updated the PR to include kubelet.service for on-prem too.

**Review by sinnykumari — APPROVED** [review]
LGTM

**Comment by sinnykumari** [comment]
This PR has got conflict due to merge of https://github.com/openshift/machine-config-operator/pull/2489

**Comment by harche** [comment]
> This PR has got conflict due to merge of #2489

Yeah, but I will also wait for https://github.com/openshift/machine-config-operator/pull/2491 to merge. Then I can just rebase it in one go.

**Comment by rphillips** [comment]
/hold

until the rebase

File: templates/common/_base/files/kubelet-auto-sizing.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
+mode: 0755
+path: "/usr/local/sbin/dynamic-system-reserved-calc.sh"
+contents:
+  inline: |
+    #!/bin/bash
+    set -e
+    NODE_SIZES_ENV=${NODE_SIZES_ENV:-/etc/node-sizing.env}
+    function dynamic_memory_sizing {
```
Comment (AlekseyUsov) [inline templates/common/_base/files/kubelet-auto-sizing.yaml:8]: What these values are based on? Not implying they're wrong, but I'm curious to see profiling data and/or graphs, if any.

File: templates/common/_base/files/kubelet-auto-sizing.yaml (hunk 3 above)
Comment (AlekseyUsov) [inline templates/common/_base/files/kubelet-auto-sizing.yaml:45]: Same question as above. Any empirical data would be useful.

**Pull request merged**